tests/test_abstract_repr.py

```diff
@@ -201,3 +216,614 @@ def test_exceptions():
+            assert wf.cls == RampWaveform
+            assert wf.kwargs["duration"] == seq_var1[0]
+            assert wf.kwargs["start"] == seq_var2[0]
+            assert wf.kwargs["stop"] == seq_var3[0]
+
+        elif wf_obj["kind"] == "blackman":
+            assert wf.cls == BlackmanWaveform
+            assert wf.kwargs["duration"] == seq_var1[0]
+            assert wf.kwargs["area"] == seq_var2[0]
+
+        elif wf_obj["kind"] == "blackman_max":
+            assert wf.cls == BlackmanWaveform.from_max_val.__wrapped__
+            assert wf.kwargs["area"] == seq_var2[0]
+            assert wf.kwargs["max_val"] == seq_var3[0]
+
+        elif wf_obj["kind"] == "interpolated":
+            assert wf.cls == InterpolatedWaveform
+            assert wf.kwargs["duration"] == seq_var1[0]
+            assert wf.kwargs["values"] == seq_var_values
+            assert wf.kwargs["times"] == seq_var_times
+
+        elif wf_obj["kind"] == "kaiser":
+            assert wf.cls == KaiserWaveform
+            assert wf.kwargs["duration"] == seq_var1[0]
+            assert wf.kwargs["area"] == seq_var3[0]
+            assert wf.kwargs["beta"] == seq_var2[0]
+
+        elif wf_obj["kind"] == "kaiser_max":
+            assert wf.cls == KaiserWaveform.from_max_val.__wrapped__
+            assert wf.kwargs["area"] == seq_var2[0]
+            assert wf.kwargs["beta"] == seq_var2[0]
+            assert wf.kwargs["max_val"] == seq_var2[0]
+
+        elif wf_obj["kind"] == "composite":
+            assert wf.cls == CompositeWaveform
+            assert all(isinstance(w, ParamObj) for w in wf.args)
+            assert all(issubclass(w.cls, Waveform) for w in wf.args)
+
+    @pytest.mark.parametrize(
+        "json_param",
+        [
+            {"expression": "neg", "lhs": {"variable": "var1"}},
+            {"expression": "abs", "lhs": {"variable": "var1"}},
+            {"expression": "ceil", "lhs": {"variable": "var1"}},
+            {"expression": "floor", "lhs": {"variable": "var1"}},
+            {"expression": "sqrt", "lhs": {"variable": "var1"}},
+            {"expression": "exp", "lhs": {"variable": "var1"}},
+            {"expression": "log", "lhs": {"variable": "var1"}},
+            {"expression": "log2", "lhs": {"variable": "var1"}},
+            {"expression": "sin", "lhs": {"variable": "var1"}},
+            {"expression": "cos", "lhs": {"variable": "var1"}},
+            {"expression": "tan", "lhs": {"variable": "var1"}},
+            {"expression": "index", "lhs": {"variable": "var1"}, "rhs": 0},
+            {"expression": "add", "lhs": {"variable": "var1"}, "rhs": 0.5},
+            {"expression": "sub", "lhs": {"variable": "var1"}, "rhs": 0.5},
+            {"expression": "mul", "lhs": {"variable": "var1"}, "rhs": 0.5},
+            {"expression": "div", "lhs": {"variable": "var1"}, "rhs": 0.5},
+            {"expression": "pow", "lhs": {"variable": "var1"}, "rhs": 0.5},
+            {"expression": "mod", "lhs": {"variable": "var1"}, "rhs": 2},
```

Comment:
I think it would be better to use the actual `var1` (ie the VariableItem rather than the array of size 1) in at least some of these.